src/Agent.Listener/Configuration/AutoLogonConfigurationManager.cs

```diff
@@ -0,0 +1,283 @@
+#if OS_WINDOWS
+using Microsoft.VisualStudio.Services.Agent.Util;
+using Microsoft.TeamFoundation.DistributedTask.WebApi;
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Security.Principal;
+using System.Threading;
+
+namespace Microsoft.VisualStudio.Services.Agent.Listener.Configuration
+{
+    [ServiceLocator(Default = typeof(AutoLogonConfigurationManager))]
+    public interface IAutoLogonConfigurationManager : IAgentService
+    {
+        void Configure(CommandSettings command);
+        void UnConfigure();
+        bool RestartNeeded();
```

Comment:
for 1: it's a tradeoff. The restart code being here bloats the already bloated configuration manager. I also would prefer this to be behind the AutoLogonConfigManager (consistent with WinSvcConfigManger). Again, the config code readability can quickly get out of hand, and we put a lot of effort into improving readability in the 2x agent. Given the readability problems present, I think moving all the auto-logon-flow behind AutoLogonConfigManager is the better trade off.

Note, regarding the outer-layer-awareness problem you mention, you've haven't solved the problem. You just pushed it back one level.